crates/key-value-spin/src/store.rs

```diff
@@ -156,6 +157,140 @@ impl Store for SqliteStore {
+            let sql_value_keys: Vec<rusqlite::types::Value> =
+                keys.into_iter().map(rusqlite::types::Value::from).collect();
+            let ptr = Rc::new(sql_value_keys);
+            self.connection
+                .lock()
+                .unwrap()
+                .prepare_cached("DELETE FROM spin_key_value WHERE store=:name AND key IN (:keys)")
+                .map_err(log_error)?
+                .execute((":name", &self.name, ":keys", ptr))
+                .map_err(log_error)
+                .map(drop)
+        })
+    }
+
+    // The assumption with increment is that if the value for the key does not exist, it will be
+    // assumed to be zero. In the case that we are unable to unmarshal the value into an i64 an error will be returned.
+    async fn increment(&self, key: String, delta: i64) -> Result<i64, Error> {
+        task::block_in_place(|| {
+            let mut binding = self.connection.lock().unwrap();
+
+            let tx = binding.transaction().map_err(log_error)?;
+
+            let value: Option<Vec<u8>> = tx
+                .prepare_cached("SELECT value FROM spin_key_value WHERE store=$1 AND key=$2")
+                .map_err(log_error)?
+                .query_map([&self.name, &key], |row| row.get(0))
+                .map_err(log_error)?
+                .next()
+                .transpose()
+                .map_err(log_error)?;
+
+            let numeric: i64 = match value {
+                Some(v) => i64::from_be_bytes(v.try_into().expect("incorrect length")),
+                None => 0,
+            };
+
+            let new_value = numeric + delta;
+            tx.prepare_cached(
+                "INSERT INTO spin_key_value (store, key, value) VALUES ($1, $2, $3)
+                     ON CONFLICT(store, key) DO UPDATE SET value=$3",
+            )
+            .map_err(log_error)?
+            .execute(rusqlite::params![&self.name, key, new_value])
+            .map_err(log_error)
+            .map(drop)?;
+
+            tx.commit().map_err(log_error)?;
+            Ok(new_value)
+        })
+    }
+
+    async fn new_compare_and_swap(&self, key: &str) -> Result<Arc<dyn Cas>, Error> {
+        let value = self.get(key).await?;
+        Ok(Arc::new(CompareAndSwap {
+            name: self.name.clone(),
+            key: key.to_string(),
+            connection: self.connection.clone(),
+            value,
+        }))
+    }
+}
+
+struct CompareAndSwap {
+    name: String,
+    key: String,
+    value: Option<Vec<u8>>,
+    connection: Arc<Mutex<Connection>>,
+}
+
+#[async_trait]
+impl Cas for CompareAndSwap {
+    async fn current(&self) -> Result<Option<Vec<u8>>, Error> {
+        Ok(self.value.clone())
+    }
+
+    async fn swap(&self, value: Vec<u8>) -> Result<(), Error> {
+        task::block_in_place(|| {
+            self.connection
+                .lock()
+                .unwrap()
+                .prepare_cached("UPDATE spin_key_value SET value=$3 WHERE store=$1 and key=$2")
```

Comment:
I think this omits the "compare" part of the CAS operation?  You might need to add a clause to the WHERE to require `value` = `self.value` (hmm, except self.value is an option).  And then check the affected row count.

The specification requires a "value has changed" CAS failure to return a `cas-failed` error containing a _new_ `cas` resource set to the current value.  Perhaps that's best done at the top level rather than repeating it in each provider, I am not sure - but the current top-level doesn't include it I think.

Reply:
I have fixed up the CAS failures and added tests to the sqlite implementation to verify the behavior.